diff --git a/include/linux/skbuff.h.orig b/include/linux/skbuff.h
index 20ca161..de8745a 100644
--- a/include/linux/skbuff.h.orig
+++ b/include/linux/skbuff.h
@@ -877,7 +877,7 @@ struct sk_buff {
 	 * want to keep them across layers you have to do a skb_clone()
 	 * first. This is owned by whoever has the skb queued ATM.
 	 */
-	char			cb[48] __aligned(8);
+	char			cb[56] __aligned(8);
 
 	union {
 		struct {
