diff --git a/include/net/netns/ipv4.h.orig b/include/net/netns/ipv4.h
index 1b80046..c891944 100644
--- a/include/net/netns/ipv4.h.orig
+++ b/include/net/netns/ipv4.h
@@ -181,6 +181,11 @@ struct netns_ipv4 {
 	unsigned int sysctl_tcp_fastopen_blackhole_timeout;
 	atomic_t tfo_active_disable_times;
 	unsigned long tfo_active_disable_stamp;
+	u8 sysctl_tcp_plb_enabled;
+	int sysctl_tcp_plb_cong_thresh;
+	u8 sysctl_tcp_plb_idle_rehash_rounds;
+	u8 sysctl_tcp_plb_rehash_rounds;
+	u8 sysctl_tcp_plb_suspend_rto_sec;
 	u32 tcp_challenge_timestamp;
 	u32 tcp_challenge_count;
 
